src/Init/Data/Cast.lean

```diff
@@ -0,0 +1,101 @@
+/-
+Copyright (c) 2014 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Gabriel Ebner
+-/
+prelude
+import Init.Data.Nat.Basic
+import Init.Data.Int.Basic
+
+/-!
+# `NatCast` and `IntCast`
+
+We introduce the typeclass `NatCast R` for a type `R` with a "canonical homomorphism" `Nat → R`.
+The typeclass carries the data of the function, but no required axioms.
+
+This typeclass was introduced to a uniform simp normal form for such morphisms.
+
+Without such a typeclass, we would have coercions such as `Int.ofNat`,
+but also later in Mathlib the generic coercion from `Nat` into any semiring (including `Int`),
+and we would need to use `simp` to move between them.
+However `simp` lemmas expressed using a non-normal form on the LHS would then not fire.
+
+In order to avoid this, we introduce this typeclass as early as possible,
+ensure it is a higher priority coercion that built-in coercions such as `Int.ofNat`,
+and write all of our simp lemmas in terms of it.
+
+Typically different instances of this class for the same target type `R` are definitionally equal,
+and so differences in the instance do not block `simp` or `rw`.
+
+The same discussion applied to `Int`, so we also introduce `IntCast`.
+
+## Note about coercions into arbitrary types:
+
+Coercions such as `Nat.cast` that go from a concrete structure such as
+`Nat` to an arbitrary type `R` should be set up as follows:
+```lean
+instance : CoeTail Nat R where coe := ...
+instance : CoeHTCT Nat R where coe := ...
+```
+
+It needs to be `CoeTail` instead of `Coe` because otherwise type-class
+inference would loop when constructing the transitive coercion `Nat → Nat → Nat → ...`.
+Sometimes we also need to declare the `CoeHTCT` instance
+if we need to shadow another coercion
+(e.g. `Nat.cast` should be used over `Int.ofNat`).
+
+-/
+
+/-- Type class for the canonical homomorphism `Nat → R`. -/
+class NatCast (R : Type u) where
+  /-- The canonical map `Nat → R`. -/
+  protected natCast : Nat → R
+
+instance : NatCast Nat where natCast n := n
+instance : NatCast Int where natCast n := Int.ofNat n
+
+/--
+Canonical homomorphism from `Nat` to a type `R`.
+
+It contains just the function, with no axioms.
+
+The prototypical example is `Int.ofNat`.
+
+In Mathlib there will be such a homomorphism whenever `R` is an additive monoid with a `1`.
```

Comment:
```suggestion
```
I don't think our docs should contain forward references to other projects/libraries - we have enough difficulties keeping them internally consistent!

Reply:
I am concerned about this general direction, referring to mathlib in lean is not / should not be illegal. We should discuss this on Zulip.

Reply:
I talked this over a bit more with @semorrison, and now I have a bit more insight into the context.

How about something like this revision that includes a bit more info?
```suggestion

This class and `IntCast` exist to allow different libraries with their own types that can be notated as natural numbers to have consistent `simp` normal forms without needing to create coercion simplification sets that are aware of all combinations. Libraries should make it easy to work with `NatCast` where possible. For instance, in Mathlib there will be such a homomorphism (and thus a `NatCast R` instance) whenever `R` is an additive monoid with a `1`.
```

Reply:
I certainly appreciate there's a problem using identifiers and filenames from downstream projects (documentation maintenance nightmare!), but that's not really an issue here. I'm pretty happy with David's rewording here.

Reply:
@digama0 We certainly don't want referring to Mathlib to be illegal! But it's important that we can provide as consistent an experience as possible and that we can flexibly move forward with important changes, making as efficient as possible use of our limited time. The goal here is not to avoid Mathlib, it's to both explain things in a way that non-Mathlib users can understand, to explain how others might use this part of the language, and to avoid fragile external references that may be broken by refactorings or API changes carried out by others. I think this reworded version addresses all these concerns, and also does a better job explaining why the code is here.